plat: Migrate rtc pl031 to drivers/ukrtc/

rares-miculescu opened the pull request:
### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Base target

  -  Architecture(s): `ARM64`
  -  Platform(s): N/A
  -  Application(s): N/A


RTC PL031 driver moved from `plat/drivers/` to `drivers/ukrtc`.

StefanJum reviewed (changes requested):
Hi @rares-miculescu, this looks good from above, few comments:

* Change the commit messages to match the [conventions](https://unikraft.org/docs/contributing/submitting-changes/#commits) (use imperative, start with capital letter `rtc moves...` -> `Move rtc code...`).
* Sign-off your commits.
* Add a `/` at the end of all directory names (`plat/drivers/` instead of `plat/driver`).

michpappas commented on drivers/ukrtc/Config.uk:
```diff
@@ -0,0 +1,4 @@
+config LIBUKRTC_PL031
+       bool "Arm platform RTC (PL031) driver"
```
With the opportunity, a better name would be "Arm PrimeCell RTC (PL031)".

michpappas commented on drivers/ukrtc/Config.uk:
```diff
@@ -0,0 +1,4 @@
+config LIBUKRTC_PL031
+       bool "Arm platform RTC (PL031) driver"
+       default y if ARCH_ARM_64
```
Since it is not a requirement that any arm-based platform has an RTC, it's best to update the corresponding platforms that feature such a device to select it instead. In fact, to give the user the possibility to disable the driver you can use `imply` instead of `select`, similarly to what I did for UART devices :

In `plat/kvm/Config.uk` you can imply PL031 in both QEMU virt and Firecracker, as they both feature such a device (you can see that from their device trees):
```
config KVM_VMM_QEMU
        bool "QEMU"
        imply LIBUKTTY_PL011

config KVM_VMM_FIRECRACKER
        bool "Firecracker"
        select KVM_BOOT_PROTO_LXBOOT
        imply LIBUKTTY_NS16550
        depends on ARCH_X86_64 || ARCH_ARM_64
```

michpappas commented on drivers/ukrtc/Makefile.uk:
```diff
@@ -0,0 +1,11 @@
+$(eval $(call addplatlib_s,kvm,libukrtc,$(CONFIG_LIBUKRTC)))
```
As discussed in #966, this can now use `addlib_s`, and the naming should be: `libukrtc_pl031` and `CONFIG_LIBUKRTC_PL031`, respectively:
```
$(eval $(call addlib_s,libukrtc_pl031,$(CONFIG_LIBUKRTC_PL031)))
```
The remaining directives must be updated too.

**Edit:** I see below you have actually done the renaming, but without the updating the registration of the library these are evalyuated to `-y` as the build system does not generate the `LIBUKRTC_PL031_CINCLUDES`.

michpappas commented on drivers/ukrtc/Makefile.uk:
```diff
@@ -0,0 +1,11 @@
+
+UK_DRIVERS_RTC_BASE := $(UK_DRIVERS_BASE)/ukrtc
+
+LIBUKRTC_PL031_CINCLUDES-y		+= -I$(LIBKVMPLAT_BASE)/include
```
Similarly to #966: You need to make public headers available via `CINCLUDES`, `ASINCLUDES`, `CXXINCLUDES`.

michpappas replied:
Please make these conditional to the library, ie substitute `LIBUKRTC_PL031_CINCLUDES-y` with  `LIBUKRTC_PL031_CINCLUDES-$(CONFIG_LIBUKRTC_PL031)` etc

michpappas commented on plat/kvm/Config.uk:
```diff
@@ -9,6 +9,7 @@ menuconfig PLAT_KVM
        select LIBFDT if ARCH_ARM_64
        select ELF64_TO_32 if ARCH_X86_64
        select LIBUKRELOC
+       select DRVUKRTC_PL031
```
Once you add `imply` to the corresponding platforms, this can go away.

michpappas commented on drivers/ukrtc/Makefile.uk:
```diff
@@ -0,0 +1,11 @@
+LIBUKRTC_PL031_CINCLUDES-y		+= -I$(LIBKVMPLAT_BASE)/include
+LIBUKRTC_PL031_CINCLUDES-y		+= -I$(UK_PLAT_COMMON_BASE)/include
+LIBUKRTC_PL031_CINCLUDES-y		+= -I$(UK_DRIVERS_RTC_BASE)/include
+LIBUKRTC_PL031_CINCLUDES-y		+= -I$(UK_PLAT_DRIVERS_BASE)/include
```
Other than `ofw`, the RTC driver has no dependencies on any code under `plat/common` or `plat/drivers`, so if you make this PR dependant on #966, the following are not needed:
```
LIBUKRTC_PL031_CINCLUDES-y		+= -I$(UK_PLAT_COMMON_BASE)/include
LIBUKRTC_PL031_CINCLUDES-y		+= -I$(UK_PLAT_COMMON_BASE)/include
LIBUKRTC_PL031_CINCLUDES-y		+= -I$(UK_PLAT_DRIVERS_BASE)/include
```

**Edit:** Actually it depends on GIC too, let's leave it as is for now, and we can see if we can merge `libofw` and `intctlr` first.

michpappas reviewed (changes requested):
Hi @rares-miculescu, besides the comments below, please update this PR with the following:
* Rebase to the latest state of #1026
* Move the driver into it's own directory, ie `drivers/ukrtc/pl031` and add a `Config.uk` under `drivers/ukrtc` with a `menu` named "Real Time Clock"
* Update no longer valid includes where needed: Now `#include <rtc/pl031.h>` must be `#include <uk/rtc/pl031.h>`. Ditto for `rtc/rtc/h`.
* Add `drivers/ukrtc/pl031exportsyms.uk` with the public symbols of this driver
* Fix conditional compilation directives where needed, Now #ifdef CONFIG_RTC_PL031` must be #if CONFIG_UKRTC_PL031`).
* Update the include guards of  the driver to `__UK_RTC_PL031_H__` and `__UK_RTC_H__` similarly to #966

Commit:
drivers: Introduce ukrtc subsystem

Add skeleton for the 'ukrtc' subsystem.	This contains the pl031 driver.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

Commit:
plat/drivers: Migrate rtc into drivers/ukrtc

Migrate the pl031 driver to the newly introduced ukrtc subsystem.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

Commit:
drivers/ukbus/platform: Add pl031 to platform devices

Add platform id for pl031. Add pl031 as compatible device for the
device-tree. The changes are made so the driver can be probed through
the platform bus.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

Commit:
drivers/ukrtc/pl031: Integrate driver for platform registration

Add implementation for platform registration. This allows the driver
to be probed by the platform bus.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

Commit:
plat/kvm: Remove initialization of PL031 from plat

Remove the initialization of PL031 from `plat/kvm/arm/setup.c`. Now
that the driver is probed by the platform bus, there is no need to be
initialized in `setup.c`.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

michpappas commented on drivers/ukrtc/pl031/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBUKRTC_PL031
+    bool "Arm PrimeCell RTC (PL031)"
```
`Config.uk` should use tab indentation

michpappas commented on drivers/ukrtc/pl031/include/uk/rtc/pl031.h:
```diff
@@ -31,13 +31,13 @@
 #define PL031_STATUS_DISABLED	0
 
-#include <rtc/rtc.h>
+#include <uk/rtc/rtc.h>
```
The current convention with headers is:
* Libraries can export a main header and optionally additional headers.
* The main header goes under `lib<name>/include/uk/<name.h>`, e.g:
  *  `lib/ukintctlr/include/uk/intctlr.h`
  * `#include <uk/intctlr.h>`
* Additional headers go under `lib<name>/include/uk/<name>/<headername>.h`
  *  `lib/ukintctlr/include/uk/intctlr/other.h`
  * `#include <uk/intctlr/other_header.h>`

Notice that this pattern is not followed by libraries introduced before re-arch (i.e. most libraries) yet, that is something we need to address as we move forward.

In `rtc` there is no API defined by a high-level library yet (i.e. `lib/ukrtc/`). Yet functionality that will be part of the future API is defined in `drivers/ukrtc/pl031/include/uk/rtc/rtc.h`. So to be forwards compatible, you may consider moving that to `drivers/ukrtc/pl031/include/uk/rtc.h` and update includes to `#include <uk/rtc.h>`.

michpappas commented on drivers/ukrtc/pl031/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBUKRTC_PL031
+    bool "Arm PrimeCell RTC (PL031)"
+    default y if ARCH_ARM_64
```
Being a peripheral, it is not necessary that all Arm cores have a PL031. What you can do instead is remove this line (remember in Kconfig `default n` is implied so you don't need to explicitly specify it), and in the Kconfig entries of of `KVM_VMM_QEMU` and `KVM_VMM_FIRECRACKER` add `imply LIBUKRTC_PL031`.

michpappas reviewed (changes requested):
@rares-miculescu only a few changes, after these I think we can merge 🚀

michpappas commented on plat/kvm/arm/setup.c:
```diff
@@ -233,11 +230,6 @@ void __no_pauth _ukplat_entry(struct ukplat_bootinfo *bi)
 
-#if CONFIG_RTC_PL031
-	/* Initialize RTC */
-	pl031_init_rtc(fdt);
```
Since now the `fdt` variable is only used by SMP you can also mark it as `__maybe_unused`, or even better you can remove it altogether and pass `(void *)bi->fdt` to `lcpu_mp_init()`.

rares-miculescu replied:
I noticed that `fdt` is passed every time to `kvm_console_init()`, therefore I don't think it should be marked as `__maybe_unused` or removed.

michpappas commented:
Also the part that refers to the formatting of commit messages in https://github.com/unikraft/unikraft/pull/1368#discussion_r1583234081 applies here too.

michpappas reviewed (approved):
Looks good, thanks @rares-miculescu 

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

razvanvirtan reviewed (approved):
All good on my side, thank you @rares-miculescu!

Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
